File: DelNet.jl
module DelNet

export blobgraph, delnetfromgraph, pushoutput!,
	   getinputs, advance!, buftostr, orderbuf


# -------------------- Data Structures --------------------
mutable struct Node
	idx_out_to_in::Int64
	num_in::Int64
	nodes_in::Array{Int64,1} 	# just for reference/construction
	idx_in_to_out::Int64
	num_out::Int64
	nodes_out::Array{Int64,1} 	#just for reference/construction
end

Node() = Node(0,0,Int64[],0,0,Int64[])


mutable struct Delay
	offset::Int64
	startidx::Int64
	len::Int64
	source::Int64
	target::Int64
end

mutable struct DelayNetwork
	inputs::Array{Float64,1}
	outputs::Array{Float64,1}
	delaybuf::Array{Float64,1}
	invidx::Array{Int64,1}
	nodes::Array{Node,1}
	delays::Array{Delay,1}
end


# -------------------- Functions --------------------
"""
"""
function pushoutput!(val, idx, dn)
	i1 = dn.nodes[idx].idx_in_to_out
	i2 = i1 + dn.nodes[idx].num_out-1 
	dn.inputs[i1:i2] .= val
end


"""
"""
function getinputs(idx, dn::DelayNetwork)
	i1 = dn.nodes[idx].idx_out_to_in
	i2 = i1 + dn.nodes[idx].num_in - 1
	dn.outputs[i1:i2] 	# note, this copies -- not a reference
end


"""
"""
function advance!(dn::DelayNetwork)
	#load input
	for i ∈ 1:length(dn.inputs)
		dn.delaybuf[ dn.delays[i].startidx + dn.delays[i].offset  ] = dn.inputs[i]	
	end

	# advance buffer
	for i ∈ 1:length(dn.inputs)
		dn.delays[i].offset = (dn.delays[i].offset + 1) % dn.delays[i].len
	end
	
	#pull output
	for i ∈ 1:length(dn.inputs)
		dn.outputs[ dn.invidx[i] ] = dn.delaybuf[ dn.delays[i].startidx + dn.delays[i].offset ]	
	end
	# output
end



"""
"""
function orderbuf(delay, delbuf) 
	[ delbuf[delay.startidx + ((delay.offset + k) % delay.len) ]
	 for k ∈ 0:delay.len-1] |> v -> map(x -> x == 0.0 ? "-" : "$(Int(round(x)))", v) |> reverse |> prod
end


"""
"""
function buftostr(buffer)
	buffer |> v -> map(x -> x == 0.0 ? "-" : "$(Int(round(x)))", v) |> prod
end


"""
"""
function blobgraph(n, p, delays::Array{Int, 1})
	@assert 0 ∉ delays "1 corresponds to no delay -- can't have 0 in delays"
	delmat = rand(n,n) |> m -> map(x -> x < p ? rand(delays) : 0, m)
	for k ∈ 1:n delmat[k,k] = 0 end
	return delmat
end


"""
"""
function delnetfromgraph(graph)
	n = size(graph)[1]
	deltot = sum(graph) 	# total number of delays to allocate
	numlines = sum( map(x -> x != 0 ? 1 : 0, graph) )
	inputs = zeros(numlines)
	outputs = zeros(numlines)
	delbuf = zeros(deltot)
	nodes_in = [[] for _ ∈ 1:n]

	nodes = [Node() for _ ∈ 1:n] 
	delays = Array{Delay, 1}(undef,numlines)


	delcount = 1
	startidx = 1
	total = 0
	for i ∈ 1:n
		for j ∈ 1:n
			if graph[i,j] != 0
				total += graph[i,j]
				nodes[i].num_out += 1
				push!(nodes_in[j], i)
				delays[delcount] = Delay(0, startidx, graph[i,j], i, j)
				startidx += graph[i,j]
				delcount += 1
			end
		end
	end

	num_outputs  = [ nd.num_out for nd ∈ nodes ]
	in_base_idcs = [ sum(num_outputs[1:k]) for k ∈ 1:n-1 ]
	in_base_idcs = [ 1; in_base_idcs[1:end] .+ 1 ]

	idx = 1
	for i ∈ 1:length(nodes)
		nodes[i].num_in = length( nodes_in[i] )
		nodes[i].idx_out_to_in = idx
		idx += nodes[i].num_in
		nodes[i].idx_in_to_out = in_base_idcs[i]
	end

	num_inputs    = [ nd.num_in for nd ∈ nodes ]
	out_base_idcs = [ sum(num_inputs[1:k]) for k ∈ 1:n-1 ]
	out_base_idcs = [ 1; out_base_idcs[1:end] .+ 1 ]
	out_counts = zeros(length(out_base_idcs))

	inverseidces = zeros(Int64, length(outputs))
	for i ∈ 1:length(inputs)
		inverseidces[i] = out_base_idcs[delays[i].target] + out_counts[delays[i].target]
		out_counts[delays[i].target] += 1
	end
		
	DelayNetwork(inputs, outputs, delbuf, inverseidces, nodes, delays)
end

end

File: delnet.c
#include <stdlib.h>
#include <stdio.h> 	//for profiling/debugging
#include <time.h> 	//for profiling/debugging

#include "delnet.h"


/*
 * -------------------- Util Functions --------------------
 */

dn_list_uint *dn_list_uint_init() {
	dn_list_uint *newlist;
	newlist = malloc(sizeof(dn_list_uint));
	newlist->count = 0;
	newlist->head = NULL; 

	return newlist;
}

void dn_list_uint_push(dn_list_uint *l, unsigned int val) {
	dn_listnode_uint *newnode;
	newnode = malloc(sizeof(dn_listnode_uint));
	newnode->val = val;
	newnode->next = l->head;
	l->head = newnode;
	l->count += 1;
}

unsigned int dn_list_uint_pop(dn_list_uint *l) {
	unsigned int val;
	dn_listnode_uint *temp;

	if (l->head != NULL) {
		val = l->head->val;
		temp = l->head;
		l->head = l->head->next;
		free(temp);
		l->count -= 1;
	}
	else {
		// trying to pop an empty list
		//val = 0;
		exit(-1);
	}
	return val;
}

void dn_list_uint_free(dn_list_uint *l) {
	while (l->head != NULL) {
		dn_list_uint_pop(l);
	}
	free(l);
}


dn_vec_float dn_orderbuf(IDX_T which, dn_delaynet *dn) {
	IDX_T k, n, idx;
	dn_vec_float output;
	
	n = dn->del_lens[which];
	output.n = n;
	output.data = malloc(sizeof(FLOAT_T) * n);

	for (k=0; k<n; k++) {
		idx = dn->del_startidces[which] +
			((dn->del_offsets[which]+k) % dn->del_lens[which]);
		output.data[n-k-1] = dn->delaybuf[idx];
	}
	return output;
}

char *dn_vectostr(dn_vec_float input) {
	int k;
	char *output;
	output = malloc(sizeof(char)*(input.n+1));
	output[input.n] = '\0';
	for(k=0; k < input.n; k++) {
		output[k] = input.data[k] == 0.0 ? '-' : '1';
	}
	return output;
}



/*
 * -------------------- delnet Functions --------------------
 */

void dn_pushoutput(FLOAT_T val, IDX_T idx, dn_delaynet *dn) 
{
	IDX_T i1, i2, k;

	i1 = dn->nodes[idx].idx_io;
	i2 = i1 + dn->nodes[idx].num_out;

	for (k = i1; k < i2; k++)
		dn->inputs[k] = val;
}


/* No getinputs()... would need to return vector */
dn_vec_float dn_getinputvec(dn_delaynet *dn) {
	dn_vec_float inputs;	
	inputs.data = malloc(sizeof(FLOAT_T)*dn->num_delays);
	inputs.n = dn->num_delays;
	for (int i=0; i<dn->num_delays; i++) {
		inputs.data[i] = dn->inputs[i];
	}
	return inputs;
}

/* get inputs to neurons (outputs of delaynet)... */
FLOAT_T *dn_getinputaddress(IDX_T idx, dn_delaynet *dn) {
	return &dn->outputs[dn->nodes[idx].idx_oi];
}


void dn_advance(dn_delaynet *dn)
{
	IDX_T k;

	for(k=0; k < dn->num_delays; k++) {
		dn->delaybuf[dn->del_startidces[k] + dn->del_offsets[k]] =
															dn->inputs[k];
	}

	for(k=0; k < dn->num_delays; k++) {
		dn->del_offsets[k] = (dn->del_offsets[k] + 1) % dn->del_lens[k];
	}

	for (k=0; k < dn->num_delays; k++) {
		dn->outputs[dn->inverseidx[k]] =
				dn->delaybuf[dn->del_startidces[k]+dn->del_offsets[k]];
	}
}


unsigned int *dn_blobgraph(unsigned int n, float p, unsigned int maxdel) {
	unsigned int count = 0;
	unsigned int *delmat;
	unsigned int i, j;
	delmat = malloc(sizeof(int)*n*n);

	for (i=0; i<n; i++) 
	for (j=0; j<n; j++) {
		if (unirand() < p && i != j) {
			delmat[i*n+j] = getrandom(maxdel) + 1;
			count += 1;
		}
		else 
			delmat[i*n+j] = 0;
	}

	return delmat;
}


dn_delaynet *dn_delnetfromgraph(unsigned int *g, unsigned int n) {
	unsigned int i, j, delcount, startidx;
	unsigned int deltot, numlines;
	dn_delaynet *dn;
	dn_list_uint **nodes_in;

	dn = malloc(sizeof(dn_delaynet));
	nodes_in = malloc(sizeof(dn_list_uint)*n);
	for (i=0; i<n; i++)
		nodes_in[i] = dn_list_uint_init();

	deltot = 0;
	numlines = 0;
	for (i=0; i<n*n; i++) {
		deltot += g[i];
		numlines += g[i] != 0 ? 1 : 0;
	}
	dn->num_delays = numlines;
	dn->buf_len = deltot;
	dn->num_nodes = n;

	dn->delaybuf = calloc(deltot, sizeof(FLOAT_T));
	dn->inputs = calloc(numlines, sizeof(FLOAT_T));
	dn->outputs = calloc(numlines, sizeof(FLOAT_T));

	dn->del_offsets = malloc(sizeof(IDX_T)*numlines);
	dn->del_startidces = malloc(sizeof(IDX_T)*numlines);
	dn->del_lens = malloc(sizeof(IDX_T)*numlines);
	dn->del_sources = malloc(sizeof(IDX_T)*numlines);
	dn->del_targets = malloc(sizeof(IDX_T)*numlines);
	dn->nodes = malloc(sizeof(dn_node)*n);

	/* init nodes */
	for (i=0; i<n; i++) {
		dn->nodes[i].idx_oi = 0;
		dn->nodes[i].num_in = 0;
		dn->nodes[i].idx_io = 0;
		dn->nodes[i].num_out = 0;
	}

	/* work through graph, allocate delay lines */
	delcount = 0;
	startidx = 0;
	for (i = 0; i<n; i++)
	for (j = 0; j<n; j++) {
		if (g[i*n + j] != 0) {
			dn_list_uint_push(nodes_in[j], i);

			dn->del_offsets[delcount] = 0;
			dn->del_startidces[delcount] = startidx;
			dn->del_lens[delcount] = g[i*n+j];
			dn->del_sources[delcount] = i;
			dn->del_targets[delcount] = j;

			dn->nodes[i].num_out += 1;

			startidx += g[i*n +j];
			delcount += 1;
		}
	}
	
	/* work out rest of index arithmetic */
	unsigned int *num_outputs, *in_base_idcs;
	num_outputs = calloc(n, sizeof(unsigned int));
	in_base_idcs = calloc(n, sizeof(unsigned int));
	for (i=0; i<n; i++) {
		num_outputs[i] = dn->nodes[i].num_out;
		for (j=0; j<i; j++)
			in_base_idcs[i] += num_outputs[j]; 	// check logic here
		//in_base_idcs[i] = i == 0 ? 0 : in_base_idcs[i-1] + num_outputs[i];
	}

	unsigned int idx = 0;
	for (i=0; i<n; i++) {
		dn->nodes[i].num_in = nodes_in[i]->count;
		dn->nodes[i].idx_oi = idx;
		idx += dn->nodes[i].num_in;
		dn->nodes[i].idx_io = in_base_idcs[i];
	}

	unsigned int *num_inputs, *out_base_idcs, *out_counts, *inverseidces;
	num_inputs = calloc(n, sizeof(unsigned int));
	out_base_idcs = calloc(n, sizeof(unsigned int));
	out_counts = calloc(n, sizeof(unsigned int));
	for (i=0; i<n; i++) {
		num_inputs[i] = dn->nodes[i].num_in;
		for (j=0; j<i; j++)
			out_base_idcs[i] += num_inputs[j]; // check logic here
		//out_base_idcs[i] = i == 0 ? 0 : in_base_idcs[i-1] + num_inputs[i];
	}

	inverseidces = calloc(numlines, sizeof(unsigned int));
	for (i=0; i < numlines; i++) {
		inverseidces[i] = out_base_idcs[dn->del_targets[i]] + 
						  out_counts[dn->del_targets[i]];
		out_counts[dn->del_targets[i]] += 1;
	}
	dn->inverseidx = inverseidces;

	/* Clean up */
	for (i=0; i<n; i++)
		dn_list_uint_free(nodes_in[i]);
	free(nodes_in);
	free(num_outputs);
	free(in_base_idcs);
	free(num_inputs);
	free(out_base_idcs);
	free(out_counts);

	return dn;
}

void dn_freedelnet(dn_delaynet *dn) {
	free(dn->del_offsets);
	free(dn->del_startidces);
	free(dn->del_lens);
	free(dn->del_sources);
	free(dn->del_targets);
	free(dn->inputs);
	free(dn->outputs);
	free(dn->inverseidx);
	free(dn->delaybuf);
	free(dn->nodes);
	free(dn);
}
